test/integration/cdi/example-driver/app/kubeletplugin.go

```diff
@@ -0,0 +1,199 @@
+	klog.Infof("Starting kubelet plugin")
+
+	// create CDI directory if not exists
+	if err := os.MkdirAll(ex.config.cdiDir, os.FileMode(0750)); err != nil {
+		return fmt.Errorf("failed to create directory: %v", err)
+	}
+
+	// create DRA driver socket directory if not exists.
+	if err := os.MkdirAll(filepath.Dir(ex.config.draAddress), 0750); err != nil {
+		return fmt.Errorf("failed to create DRA driver socket directory: %v", err)
+	}
+
+	// Run a grpc server for the driver, that listens on draAddress
+	if err := plugin.StartNonblockingGrpcServer(ex.config.draAddress, &ex); err != nil {
+		return fmt.Errorf("failed to start a nonblocking grpc server: %v", err)
+	}
+
+	// Run a registration server and registers plugin with kubelet
+	if err := plugin.StartRegistrar(ex.config.driverName, ex.config.draAddress, ex.config.pluginRegistrationPath); err != nil {
+		return fmt.Errorf("failed to start registrar: %v", err)
+	}
+
+	return nil
+
+}
+
+func (ex *examplePlugin) NodePrepareResource(ctx context.Context, req *drapbv1.NodePrepareResourceRequest) (*drapbv1.NodePrepareResourceResponse, error) {
+
+	const (
+		deviceName = "exampledevice"
+		cdiVersion = "0.2.0"
+	)
+	kind := ex.config.driverName + "/" + deviceName
+
+	klog.Infof("NodePrepareResource is called: request: %+v", req)
+	// create CDI Files
+
+	// create spec
+	klog.Infof("Creating CDI Files")
+
+	// resourceHandle in the form of map[string]string
+	var decodedResourceHandle map[string]string
+	if err := json.Unmarshal([]byte(req.ResourceHandle), &decodedResourceHandle); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal resourceHandle: %v", err)
+	}
+	// validate if resourceHandle
+	if err := ex.validateResourceHandle(decodedResourceHandle); err != nil {
+		return nil, fmt.Errorf("failed to validate resourceHandle: %v", err)
+	}
+
+	containerEditsInDevices := (&cdiapi.ContainerEdits{}).Append(&cdiapi.ContainerEdits{ContainerEdits: &specs.ContainerEdits{}})
+	for key, val := range decodedResourceHandle {
+		containerEditsInDevices.Append(&cdiapi.ContainerEdits{ContainerEdits: &specs.ContainerEdits{
+			Env: []string{key + "=" + val},
+		}})
+	}
+
+	containerEdits := (&cdiapi.ContainerEdits{}).Append(&cdiapi.ContainerEdits{ContainerEdits: &specs.ContainerEdits{}})
+
+	spec, err := cdiapi.NewSpec(
+		&specs.Spec{
+			Version: cdiVersion,
+			Kind:    kind,
+			Devices: []specs.Device{{
+				Name:           deviceName,
+				ContainerEdits: *containerEditsInDevices.ContainerEdits,
+			}},
+			ContainerEdits: *containerEdits.ContainerEdits,
+		},
+		ex.getJsonFilePath(req.ClaimUid), 1)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new spec: %v", err)
+	}
+	klog.V(5).InfoS("CDI spec is validated")
+
+	// create json file from spec
+	doc, err := json.Marshal(*spec)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal: %v", err)
+	}
+
+	if err := os.WriteFile(ex.getJsonFilePath(req.ClaimUid), doc, os.FileMode(0644)); err != nil {
```

Comment:
It was suggested by CDI devs to use [Tempfile](https://pkg.go.dev/io/ioutil#TempFile) to avoid name conflicts. Please, use that.

Reply:
Thanks!

Reply:
yep, use os.CreateTemp or any API that creates temporary file.

Reply:
Why was this marked as resolved? It's still using `os.WriteFile`.

Regarding the proposal to use some tempfile function: that makes it harder to ensure that the function is idempotent. Before writing a new temp file, it would have to read all CDI files to check whether a previous invocation already wrote it.

I prefer to have a deterministic name that is guaranteed (by convention) to be unique. If no such convention exists, then CDI should define one.

Reply:
makes sense to me.

Reply:
@pohly Because,, you are writing hundreds of comments, I cannot see what is what if I do not just resolve after solving in my local env.

Do you mean, we need to still use writefile?
getJsonFilePath will return the name you suggested earlier.

Reply:
> Because,, you are writing hundreds of comments

Just pick up one conversation, fix it. Push the code and only after that resolve. You don't need to resolve the all at once. Alternatively we can agree that reviewers will resolve them when they think they are resolved.

Reply:
Ok, from now on, I will never click 'Resolve'. Either of you resolve please!

Reply:
agreed